src/OpenLoco/Company.cpp

```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
     // 0x00430762
     void Company::aiThink()
     {
-        registers regs;
-        regs.esi = X86Pointer(this);
-        call(0x00430762, regs);
+        const auto updatingCompanyId = CompanyManager::updatingCompanyId();
+
+        // Ensure this is only used for Non-Player controlled companies.
+        if (CompanyManager::isPlayerCompany(updatingCompanyId))
+            return;
+
+        const auto thinkFunc1 = _funcs_430786[var_4A4];
+        thinkFunc1(this);
+
+        if (empty())
+            return;
+
+        const auto thinkFunc2 = _funcs_43079E[var_4A6];
+        thinkFunc2(this);
+
+        if (headquarters_x == -1 && (challenge_flags & CompanyFlags::bankrupt) == 0 && (challenge_flags & 0x1) != 0)
+        {
+            // Look for an entry with either town or industry assigned.
+            auto index = std::size(var_4A8);
+            while (var_4A8[--index].var_00 == 0xFF)
+            {
+                if (index == 0)
+                    return;
+            }
+
+            auto& entry = var_4A8[index];
+
+            Map::Pos2 pos;
+            if ((_dword4FE720[entry.var_00] & 2) != 0)
+            {
+                auto* industry = IndustryManager::get(static_cast<IndustryId>(entry.var_01));
+                pos = { industry->x, industry->y };
+            }
+            else
+            {
+                auto* town = TownManager::get(static_cast<TownId>(entry.var_01));
+                pos = { town->x, town->y };
+            }
+
+            auto& prng = gPrng();
+            const auto randPick = prng.randNext();
+            const auto randPos = Map::Pos2{
+                static_cast<coord_t>(randPick & 0x3E0),
+                static_cast<coord_t>(Utility::ror<uint32_t>(randPick, 5) & 0x3E0)
+            };
+            const auto selectedPos = randPos - Map::Pos2{ 512, 512 } + pos;
+            if (selectedPos.x <= 0x2FFF && selectedPos.y <= 0x2FFF)
+            {
+                auto tile = Map::TileManager::get(selectedPos);
+                auto* surface = tile.surface();
+
+                coord_t z = surface->baseZ() << 2;
+                if (surface->slope() != 0)
+                    z += 16;
+
+                const auto rot = randPos.y & 3;
```

Comment:
This wont work. randPos.y has been &0x3E0. You need to get the rand out of `randPick`.
rot = (Utility::ror<uint32_t>(randPick, 10)) & 0x3;